### PR TITLE
Better docs in IOUtils and IOUtils.byteArray(int size)

### PR DESCRIPTION
**Pull request by ArdenL-Liu**
IOUtils.byteArray(int size) add the verification to assure that the size is legal(size > 0), the illegal(size <=0) should throw IllegalArgumentException.

### CODE REVIEW

File: src/test/java/org/apache/commons/io/IOUtilsTest.java [hunk 1]
```diff
@@ -1716,4 +1716,18 @@ public void testWriteLittleString() throws IOException {
...
+    public void testByteArrayWithIllegalSize() {
+        try {
+            int size = -1;
+            byte[] bytes = IOUtils.byteArray(size);
+
+            size = 0;
+            bytes = IOUtils.byteArray(size);
+        }catch (Exception e) {
+            assertEquals(e.getClass().getName(), IllegalArgumentException.class.getName());
+        }
```
Comment (kinow) [inline src/test/java/org/apache/commons/io/IOUtilsTest.java:1729]: I think instead it could use JUnit's `@Test(expected=IllegalArgumentException.class)`, or `assertThrows`.

Comment (kinow) [reply src/test/java/org/apache/commons/io/IOUtilsTest.java:1729]: And I think when the first exception happens, the second is part of the test is never executed? Probably needs to be broken down into two separate tests, maybe with `@ParameterizedTest` too.

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 2]
```diff
@@ -346,9 +346,15 @@ public static byte[] byteArray() {
      *
      * @param size array size.
      * @return a new byte array of the given size.
+     *
+     * @exception  IllegalArgumentException  If {@code size <= 0}
```
Comment (kinow) [inline src/main/java/org/apache/commons/io/IOUtils.java:350]: `@throws` instead of `@exception`

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 3]
```diff
@@ -3724,7 +3724,9 @@ public static Writer writer(final Appendable appendable) {
 
     /**
      * Instances should NOT be constructed in standard programming.
+     * @deprecated Will be private in 3.0.
      */
+    @Deprecated
```
Comment (kinow) [inline src/main/java/org/apache/commons/io/IOUtils.java:3729]: This is doing more than what's described in the PR title. Maybe worth moving to a separate PR to make it easier to track in the changelog.

**Review by kinow — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/commons/io/IOUtils.java (hunk 3 above)
Comment (ArdenL-Liu) [reply src/main/java/org/apache/commons/io/IOUtils.java:3729]: Thx, I will rollback it

File: src/main/java/org/apache/commons/io/IOUtils.java (hunk 2 above)
Comment (ArdenL-Liu) [reply src/main/java/org/apache/commons/io/IOUtils.java]: Thx, update it with @throws

File: src/test/java/org/apache/commons/io/IOUtilsTest.java (hunk 1 above)
Comment (ArdenL-Liu) [reply src/test/java/org/apache/commons/io/IOUtilsTest.java]: ok, I changed IOUtilsTest.testByteArrayWithIllegalSize() with @ParameterizedTest.
you can review the code.

File: src/main/java/org/apache/commons/io/FileUtils.java [hunk 4]
```diff
@@ -3537,9 +3537,7 @@ public static void writeStringToFile(final File file, final String data, final S
 
     /**
      * Instances should NOT be constructed in standard programming.
-     * @deprecated Will be private in 3.0.
      */
-    @Deprecated
```
Comment (pjfanning) [inline src/main/java/org/apache/commons/io/FileUtils.java:3542]: Should this `@Deprecated` be removed? There was a comment to remove a different `@Deprecated` but this one should be kept.

File: src/test/java/org/apache/commons/io/IOUtilsTest.java (hunk 1 above)
Comment (garydgregory) [reply src/test/java/org/apache/commons/io/IOUtilsTest.java]: Why would we trade a more vague exception for the more precise and quite specific one already thrown by the VM? Maybe this issue can be simplified with better Javadoc?

File: src/main/java/org/apache/commons/io/FileUtils.java (hunk 4 above)
Comment (ArdenL-Liu) [reply src/main/java/org/apache/commons/io/FileUtils.java:3542]: rollback it

File: src/test/java/org/apache/commons/io/IOUtilsTest.java (hunk 1 above)
Comment (ArdenL-Liu) [reply src/test/java/org/apache/commons/io/IOUtilsTest.java]: Hi garydgregory,
    Thank you for your good suggestion, I updated the doc and changed the unittest, you can review it again.

File: src/main/java/org/apache/commons/io/FileUtils.java (hunk 4 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/io/FileUtils.java:3542]: -1: Don't remove this tag,

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 5]
```diff
@@ -341,6 +341,7 @@ public static byte[] byteArray() {
 
     /**
      * Returns a new byte array of the given size.
+     * Throws java.lang.NegativeArraySizeException if the size is negative.
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/IOUtils.java:344]: This is in the wrong place, use a throws tag.

File: src/test/java/org/apache/commons/io/IOUtilsTest.java [hunk 6]
```diff
@@ -1716,4 +1718,12 @@ public void testWriteLittleString() throws IOException {
         }
     }
 
+    @Test
+    public void testByteArrayWithNegativeSize() {
+        int size = -1;
+        assertThrows(NegativeArraySizeException.class,() -> {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/IOUtilsTest.java:1724]: You can just say `assertThrows(NegativeArraySizeException.class,() -> IOUtils.byteArray(-1));` but I am not even sure we should have this test since it does not improve our code coverage, it just test what the JVM does.

File: src/test/java/org/apache/commons/io/IOUtilsTest.java [hunk 7]
```diff
@@ -79,6 +79,8 @@
 import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
+import org.junit.jupiter.params.ParameterizedTest;
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/IOUtilsTest.java:82]: -1: Unused

**Review by garydgregory — CHANGES_REQUESTED** [review]
Hello @ArdenL-Liu 
Thank you for the PR, please see my comments.

File: src/test/java/org/apache/commons/io/IOUtilsTest.java (hunk 6 above)
Comment (ArdenL-Liu) [reply src/test/java/org/apache/commons/io/IOUtilsTest.java:1683]: Hi garydgregory,
    from your comments, I checked the code and fixed again, you can review it again.
Thanks.

File: src/main/java/org/apache/commons/io/IOUtils.java (hunk 5 above)
Comment (ArdenL-Liu) [reply src/main/java/org/apache/commons/io/IOUtils.java]: Hi garydgregory,
   I fixed it with @throws, you can review it again.

File: src/main/java/org/apache/commons/io/FileUtils.java (hunk 4 above)
Comment (pjfanning) [reply src/main/java/org/apache/commons/io/FileUtils.java:3542]: this is not resolved - you have still removed a comment that was in the code before - can you put it back?

File: src/main/java/org/apache/commons/io/IOUtils.java (hunk 3 above)
Comment (pjfanning) [reply src/main/java/org/apache/commons/io/IOUtils.java:3724]: Can you remove this deprecation as @kinow requested?

File: src/main/java/org/apache/commons/io/FileUtils.java (hunk 4 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/io/FileUtils.java:3542]: The `@deprecated` tag should NOT be removed.

Comment (ArdenL-Liu) [reply src/main/java/org/apache/commons/io/FileUtils.java]: Hi All,
     sorry for my error operation, I fixed the doc and add @deprecated tag about FileUtils.

File: src/test/java/org/apache/commons/io/IOUtilsTest.java (hunk 1 above)
Comment (ArdenL-Liu) [reply src/test/java/org/apache/commons/io/IOUtilsTest.java]: In order to improve my work efficiency，I will study github workflow in the next week.

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 8]
```diff
@@ -346,6 +346,7 @@ public static byte[] byteArray() {
      *
      * @param size array size.
      * @return a new byte array of the given size.
+     * @throws NegativeArraySizeException if the size is negative.
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/IOUtils.java:349]: This PR seems ok to me now except for this addition. We do not document this type of JVM exception anywhere else IIRC. If we were to doc this, then it would make sense to doc all callers as well, then this would touch almost everything which would not help anyone IMO.

File: src/test/java/org/apache/commons/io/IOUtilsTest.java [hunk 9]
```diff
@@ -16,40 +16,20 @@
  */
 package org.apache.commons.io;
 
-import static org.junit.jupiter.api.Assertions.assertArrayEquals;
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/IOUtilsTest.java:19]: Remove noise in the PR: There is no need to update these imports.

File: src/test/java/org/apache/commons/io/IOUtilsTest.java [hunk 10]
```diff
@@ -60,25 +40,7 @@
...
-import org.apache.commons.io.output.NullWriter;
-import org.apache.commons.io.output.StringBuilderWriter;
-import org.apache.commons.io.test.TestUtils;
-import org.apache.commons.io.test.ThrowOnCloseReader;
-import org.apache.commons.lang3.StringUtils;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Disabled;
-import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.io.TempDir;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/IOUtilsTest.java:43]: "*" imports are a no go

Comment (ArdenL-Liu) [reply src/test/java/org/apache/commons/io/IOUtilsTest.java]: fixed

File: src/main/java/org/apache/commons/io/IOUtils.java (hunk 8 above)
Comment (ArdenL-Liu) [reply src/main/java/org/apache/commons/io/IOUtils.java:349]: As yours, this is why I want to add throws NegativeArraySizeException tag to this method,  In fact, in order to help  users to use and understand these methods,  I think we should add it to the doc of  all methods like it.

**Comment by garydgregory** [comment]
FYI: I've handled the Javadoc "java.lang" issue separately throughout most Apache Commons components.

**Comment by garydgregory** [comment]
@ArdenL-Liu
Please **rebase** on git master and see any comments that may be outstanding.

**Comment by ArdenL-Liu** [comment]
Hi garydgregory，
  thank you, I will rebase on git master.

**Comment by ArdenL-Liu** [comment]
Hi garydgregory，
    You're great.  the comments updated in master line.

File: src/main/java/org/apache/commons/io/IOUtils.java (hunk 3 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/io/IOUtils.java:3724]: Wait... this deprecation is fine. It's slightly out of scope for the PR but it's fine.
